File: Project Academy/11. Pagination/java/app/src/main/java/com/dicoding/academies/data/source/local/LocalDataSource.java
package com.dicoding.academies.data.source.local;

import androidx.lifecycle.LiveData;
import androidx.paging.DataSource;

import com.dicoding.academies.data.source.local.entity.CourseEntity;
import com.dicoding.academies.data.source.local.entity.CourseWithModule;
import com.dicoding.academies.data.source.local.entity.ModuleEntity;
import com.dicoding.academies.data.source.local.room.AcademyDao;

import java.util.List;

public class LocalDataSource {

    private static LocalDataSource INSTANCE;
    private final AcademyDao mAcademyDao;

    private LocalDataSource(AcademyDao mAcademyDao) {
        this.mAcademyDao = mAcademyDao;
    }

    public static LocalDataSource getInstance(AcademyDao academyDao) {
        if (INSTANCE == null) {
            INSTANCE = new LocalDataSource(academyDao);
        }
        return INSTANCE;
    }

    public DataSource.Factory<Integer, CourseEntity> getAllCourses() {
        return mAcademyDao.getCourses();
    }

    public DataSource.Factory<Integer, CourseEntity> getBookmarkedCourses() {
        return mAcademyDao.getBookmarkedCourse();
    }

    public LiveData<CourseWithModule> getCourseWithModules(final String courseId) {
        return mAcademyDao.getCourseWithModuleById(courseId);
    }

    public LiveData<List<ModuleEntity>> getAllModulesByCourse(String courseId) {
        return mAcademyDao.getModulesByCourseId(courseId);
    }

    public void insertCourses(List<CourseEntity> courses) {
        mAcademyDao.insertCourses(courses);
    }

    public void insertModules(List<ModuleEntity> modules) {
        mAcademyDao.insertModules(modules);
    }

    public void setCourseBookmark(CourseEntity course, boolean newState) {
        course.setBookmarked(newState);
        mAcademyDao.updateCourse(course);
    }

    public LiveData<ModuleEntity> getModuleWithContent(String moduleId) {
        return mAcademyDao.getModuleById(moduleId);
    }

    public void updateContent(String content, String moduleId) {
        mAcademyDao.updateModuleByContent(content, moduleId);
    }

    public void setReadModule(final ModuleEntity module) {
        module.setRead(true);
        mAcademyDao.updateModule(module);
    }
}

File: Project Academy/11. Pagination/kotlin/app/src/main/java/com/dicoding/academies/data/source/local/LocalDataSource.kt
package com.dicoding.academies.data.source.local

import androidx.lifecycle.LiveData
import androidx.paging.DataSource

import com.dicoding.academies.data.source.local.entity.CourseEntity
import com.dicoding.academies.data.source.local.entity.CourseWithModule
import com.dicoding.academies.data.source.local.entity.ModuleEntity
import com.dicoding.academies.data.source.local.room.AcademyDao

class LocalDataSource private constructor(private val mAcademyDao: AcademyDao) {

    companion object {
        private var INSTANCE: LocalDataSource? = null

        fun getInstance(academyDao: AcademyDao): LocalDataSource {
            return INSTANCE ?: LocalDataSource(academyDao)
        }
    }

    fun getAllCourses(): DataSource.Factory<Int, CourseEntity> = mAcademyDao.getCourses()

    fun getBookmarkedCourses(): DataSource.Factory<Int, CourseEntity> =
            mAcademyDao.getBookmarkedCourse()

    fun getCourseWithModules(courseId: String): LiveData<CourseWithModule> =
            mAcademyDao.getCourseWithModuleById(courseId)

    fun getAllModulesByCourse(courseId: String): LiveData<List<ModuleEntity>> =
            mAcademyDao.getModulesByCourseId(courseId)

    fun insertCourses(courses: List<CourseEntity>) {
        mAcademyDao.insertCourses(courses)
    }

    fun insertModules(modules: List<ModuleEntity>) {
        mAcademyDao.insertModules(modules)
    }

    fun setCourseBookmark(course: CourseEntity, newState: Boolean) {
        course.bookmarked = newState
        mAcademyDao.updateCourse(course)
    }

    fun getModuleWithContent(moduleId: String): LiveData<ModuleEntity> =
            mAcademyDao.getModuleById(moduleId)

    fun updateContent(content: String, moduleId: String) {
        mAcademyDao.updateModuleByContent(content, moduleId)
    }

    fun setReadModule(module: ModuleEntity) {
        module.read = true
        mAcademyDao.updateModule(module)
    }
}